packages/react-instantsearch-hooks/src/lib/useInstantSearchApi.ts

```diff
@@ -84,8 +66,102 @@ export function useInstantSearchApi<TUiState extends UiState, TRouteState>(
       // the server state and pass it to the render on SSR.
       serverContext.notifyServer({ search });
     }
+
+    searchRef.current = search;
   }
 
+  {
+    const search = searchRef.current;
+
+    if (prevProps.indexName !== props.indexName) {
+      search.helper!.setIndex(props.indexName).search();
+      setPrevProps(props);
+    }
+
+    if (prevProps.searchClient !== props.searchClient) {
+      addAlgoliaAgents(props.searchClient, [
+        ...defaultUserAgents,
+        serverContext && serverUserAgent,
+      ]);
+      search.mainHelper!.setClient(props.searchClient).search();
+      setPrevProps(props);
+    }
+
+    if (prevProps.onStateChange !== props.onStateChange) {
+      search.onStateChange = props.onStateChange;
+      setPrevProps(props);
+    }
+
+    if (prevProps.searchFunction !== props.searchFunction) {
+      // Updating the `searchFunction` to `undefined` is not supported by
+      // InstantSearch.js, so it will throw an error.
+      // This is a fair behavior until we add an update API in InstantSearch.js.
+      search._searchFunction = props.searchFunction;
+      setPrevProps(props);
+    }
+
+    if (prevProps.stalledSearchDelay !== props.stalledSearchDelay) {
+      // The default `stalledSearchDelay` in InstantSearch.js is 200ms.
+      // We need to reset it when it's undefined to get back to the original value.
+      search._stalledSearchDelay = props.stalledSearchDelay ?? 200;
+      setPrevProps(props);
+    }
+
+    // Updating the `routing` prop is not supported because InstantSearch.js
+    // doesn't let us change it. This might not be a problem though, because `routing`
+    // shouldn't need to be dynamic.
+    // If we find scenarios where `routing` needs to change, we can always expose
+    // it privately on the InstantSearch instance. Another way would be to
+    // manually inject the routing middleware in this library, and not rely
+    // on the provided `routing` prop.
```

Comment:
still not sure whether we should let people know routing will be stale, in theory if the stale routing is the same as the newer ones (but just not referentially) there shouldn't be an issue, but some weird edge cases like using react state in the router would break